Description: Backport CVE-2026-41716 (GHSA-9fw2-h3hf-293r) to spring-data-commons 3.3.5.
Adapted from OSS 3.5.12 commits e71100a8b, 239860738, e33b68517. https://slim-ai.atlassian.net/browse/TP-142219
Origin: adapted, CVE-2026-41716

--- a/.mvn/extensions.xml
+++ b/.mvn/extensions.xml
@@ -1,8 +1,3 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <extensions>
-	<extension>
-		<groupId>io.spring.develocity.conventions</groupId>
-		<artifactId>develocity-conventions-maven-extension</artifactId>
-		<version>0.0.19</version>
-	</extension>
 </extensions>
--- a/src/main/java/org/springframework/data/convert/DefaultTypeMapper.java
+++ b/src/main/java/org/springframework/data/convert/DefaultTypeMapper.java
@@ -18,10 +18,7 @@
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map;
 import java.util.Optional;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Function;
 
 import org.springframework.beans.factory.BeanClassLoaderAware;
 import org.springframework.data.mapping.Alias;
@@ -30,6 +27,7 @@
 import org.springframework.data.util.TypeInformation;
 import org.springframework.lang.Nullable;
 import org.springframework.util.Assert;
+import org.springframework.util.ConcurrentLruCache;
 
 /**
  * Default implementation of {@link TypeMapper}.
@@ -43,9 +41,7 @@
 
 	private final TypeAliasAccessor<S> accessor;
 	private final List<? extends TypeInformationMapper> mappers;
-	private final Map<Alias, Optional<TypeInformation<?>>> typeCache;
-
-	private final Function<Alias, Optional<TypeInformation<?>>> getAlias;
+	private final ConcurrentLruCache<Alias, Optional<TypeInformation<?>>> typeCache;
 
 	/**
 	 * Creates a new {@link DefaultTypeMapper} using the given {@link TypeAliasAccessor}. It will use a
@@ -69,7 +65,7 @@
 	}
 
 	/**
-	 * Creates a new {@link DefaultTypeMapper} using the given {@link TypeAliasAccessor}, {@link MappingContext} and
+	 * Creates a new {@link DefaultTypeMapper} with a cache size of 1024, using the given {@link TypeAliasAccessor}, {@link MappingContext} and
 	 * additional {@link TypeInformationMapper}s. Will register a {@link MappingContextTypeInformationMapper} before the
 	 * given additional mappers.
 	 *
@@ -80,6 +76,23 @@
 	public DefaultTypeMapper(TypeAliasAccessor<S> accessor,
 			@Nullable MappingContext<? extends PersistentEntity<?, ?>, ?> mappingContext,
 			List<? extends TypeInformationMapper> additionalMappers) {
+		this(accessor, mappingContext, additionalMappers, 1024);
+	}
+
+	/**
+	 * Creates a new {@link DefaultTypeMapper} with the given cache size, using the given {@link TypeAliasAccessor}, {@link MappingContext} and
+	 * additional {@link TypeInformationMapper}s. Will register a {@link MappingContextTypeInformationMapper} before the
+	 * given additional mappers.
+	 *
+	 * @param accessor must not be {@literal null}.
+	 * @param mappingContext
+	 * @param additionalMappers must not be {@literal null}.
+	 * @param cacheSize must not be negative.
+	 * @since 3.5.12
+	 */
+	public DefaultTypeMapper(TypeAliasAccessor<S> accessor,
+		@Nullable MappingContext<? extends PersistentEntity<?, ?>, ?> mappingContext,
+		List<? extends TypeInformationMapper> additionalMappers, int cacheSize) {
 
 		Assert.notNull(accessor, "Accessor must not be null");
 		Assert.notNull(additionalMappers, "AdditionalMappers must not be null");
@@ -92,18 +105,7 @@
 
 		this.mappers = Collections.unmodifiableList(mappers);
 		this.accessor = accessor;
-		this.typeCache = new ConcurrentHashMap<>();
-		this.getAlias = key -> {
-
-			for (TypeInformationMapper mapper : mappers) {
-				TypeInformation<?> typeInformation = mapper.resolveTypeFrom(key);
-
-				if (typeInformation != null) {
-					return Optional.of(typeInformation);
-				}
-			}
-			return Optional.empty();
-		};
+		this.typeCache = new ConcurrentLruCache<>(cacheSize, this::resolveTypeInformation);
 	}
 
 	@Nullable
@@ -116,6 +118,25 @@
 	}
 
 	/**
+	 * Tries to lookup a {@link TypeInformation} for the given alias in the {@link #mappers} and return the first one found.
+	 *
+	 * @param alias must not be {@literal null}.
+	 * @return the {@link TypeInformation} for the given alias or {@link Optional#empty()} if none was found.
+	 * @since 3.5.12
+	 */
+	private Optional<TypeInformation<?>> resolveTypeInformation(Alias alias) {
+
+		for (TypeInformationMapper mapper : mappers) {
+			TypeInformation<?> typeInformation = mapper.resolveTypeFrom(alias);
+
+			if (typeInformation != null) {
+				return Optional.of(typeInformation);
+			}
+		}
+		return Optional.empty();
+	}
+
+	/**
 	 * Tries to lookup a {@link TypeInformation} for the given alias from the cache and return it if found. If none is
 	 * found it'll consult the {@link TypeInformationMapper}s and cache the value found.
 	 *
@@ -124,7 +145,7 @@
 	 */
 	@Nullable
 	private TypeInformation<?> getFromCacheOrCreate(Alias alias) {
-		return typeCache.computeIfAbsent(alias, getAlias).orElse(null);
+		return typeCache.get(alias).orElse(null);
 	}
 
 	@Override
--- a/src/main/java/org/springframework/data/convert/SimpleTypeInformationMapper.java
+++ b/src/main/java/org/springframework/data/convert/SimpleTypeInformationMapper.java
@@ -15,15 +15,14 @@
  */
 package org.springframework.data.convert;
 
-import java.util.Map;
 import java.util.Optional;
-import java.util.concurrent.ConcurrentHashMap;
 
 import org.springframework.beans.factory.BeanClassLoaderAware;
 import org.springframework.data.mapping.Alias;
 import org.springframework.data.util.TypeInformation;
 import org.springframework.lang.Nullable;
 import org.springframework.util.ClassUtils;
+import org.springframework.util.ConcurrentLruCache;
 
 /**
  * Basic {@link TypeInformationMapper} implementation that interprets the alias handles as fully qualified class name
@@ -32,14 +31,32 @@
  *
  * @author Oliver Gierke
  * @author Mark Paluch
+ * @author Christoph Strobl
  */
 public class SimpleTypeInformationMapper implements TypeInformationMapper, BeanClassLoaderAware {
 
-	private final Map<String, Optional<TypeInformation<?>>> cache = new ConcurrentHashMap<>();
+	private final ConcurrentLruCache<String, Optional<TypeInformation<?>>> cache;
 
 	private @Nullable ClassLoader classLoader;
 
 	/**
+	 * Creates a new {@link SimpleTypeInformationMapper} with a cache size of 1024.
+	 */
+	public SimpleTypeInformationMapper() {
+		this(1024);
+	}
+
+	/**
+	 * Creates a new {@link SimpleTypeInformationMapper} with the given cache size.
+	 *
+	 * @param cacheSize must not be negative.
+	 * @since 3.5.12
+	 */
+	public SimpleTypeInformationMapper(int cacheSize) {
+		this.cache = new ConcurrentLruCache<>(cacheSize, this::loadClass);
+	}
+
+	/**
 	 * Returns the {@link TypeInformation} that shall be used when the given {@link String} value is found as type hint.
 	 * The implementation will simply interpret the given value as fully-qualified class name and try to load the class.
 	 * Will return {@literal null} in case the given {@link String} is empty.
@@ -55,7 +72,7 @@
 		String stringAlias = alias.mapTyped(String.class);
 
 		if (stringAlias != null) {
-			return cache.computeIfAbsent(stringAlias, this::loadClass).orElse(null);
+			return cache.get(stringAlias).orElse(null);
 		}
 
 		return null;
--- a/src/main/java/org/springframework/data/mapping/PropertyPath.java
+++ b/src/main/java/org/springframework/data/mapping/PropertyPath.java
@@ -45,6 +45,7 @@
 public class PropertyPath implements Streamable<PropertyPath> {
 
 	private static final String PARSE_DEPTH_EXCEEDED = "Trying to parse a path with depth greater than 1000; This has been disabled for security reasons to prevent parsing overflows";
+	private static final int MAX_PARSE_DEPTH = 1000;
 
 	private static final String DELIMITERS = "_\\.";
 	private static final Pattern SPLITTER = Pattern.compile("(?:[%s]?([%s]*?[^%s]+))".replaceAll("%s", DELIMITERS));
@@ -371,7 +372,11 @@
 					: SPLITTER.matcher("_" + it.path);
 
 			while (matcher.find()) {
+
 				iteratorSource.add(matcher.group(1));
+				if(iteratorSource.size() > MAX_PARSE_DEPTH) {
+					throw new IllegalArgumentException(PARSE_DEPTH_EXCEEDED);
+				}
 			}
 
 			Iterator<String> parts = iteratorSource.iterator();
@@ -428,7 +433,7 @@
 	 * @return
 	 */
 	private static PropertyPath create(String source, TypeInformation<?> type, List<PropertyPath> base) {
-		return create(source, type, "", base);
+		return create(source, type, "", base, MAX_PARSE_DEPTH - base.size());
 	}
 
 	/**
@@ -441,9 +446,9 @@
 	 * @param addTail
 	 * @return
 	 */
-	private static PropertyPath create(String source, TypeInformation<?> type, String addTail, List<PropertyPath> base) {
+	private static PropertyPath create(String source, TypeInformation<?> type, String addTail, List<PropertyPath> base, int remainingDepth) {
 
-		if (base.size() > 1000) {
+		if (base.size() > MAX_PARSE_DEPTH || remainingDepth <= 0) {
 			throw new IllegalArgumentException(PARSE_DEPTH_EXCEEDED);
 		}
 
@@ -485,7 +490,7 @@
 			String tail = source.substring(position);
 
 			try {
-				return create(head, type, tail + addTail, base);
+				return create(head, type, tail + addTail, base, remainingDepth - 1);
 			} catch (PropertyReferenceException e) {
 				throw e.hasDeeperResolutionDepthThan(exception) ? e : exception;
 			}
--- a/src/main/java/org/springframework/data/util/TypeDiscoverer.java
+++ b/src/main/java/org/springframework/data/util/TypeDiscoverer.java
@@ -22,9 +22,9 @@
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
@@ -37,6 +37,7 @@
 import org.springframework.util.ConcurrentLruCache;
 import org.springframework.util.ObjectUtils;
 import org.springframework.util.ReflectionUtils;
+import org.springframework.util.ReflectionUtils.FieldCallback;
 
 /**
  * Basic {@link TypeDiscoverer} that contains basic functionality to discover property types.
@@ -54,7 +55,7 @@
 			TypeDiscoverer::new);
 
 	private final ResolvableType resolvableType;
-	private final Map<String, Optional<TypeInformation<?>>> fields = new ConcurrentHashMap<>();
+	private final Lazy<Map<String, TypeInformation<?>>> fields;
 	private final Lazy<TypeInformation<?>> componentType;
 	private final Lazy<TypeInformation<?>> valueType;
 	private final Map<Constructor<?>, List<TypeInformation<?>>> constructorParameters = new ConcurrentHashMap<>();
@@ -67,6 +68,7 @@
 		Assert.notNull(type, "Type must not be null");
 
 		this.resolvableType = type;
+		this.fields = Lazy.of(this::doGetProperties);
 		this.componentType = Lazy.of(this::doGetComponentType);
 		this.valueType = Lazy.of(this::doGetMapValueType);
 		this.typeArguments = Lazy.of(this::doGetTypeArguments);
@@ -107,7 +109,7 @@
 		var separatorIndex = name.indexOf('.');
 
 		if (separatorIndex == -1) {
-			return fields.computeIfAbsent(name, this::getPropertyInformation).orElse(null);
+			return fields.get().get(name);
 		}
 
 		var head = name.substring(0, separatorIndex);
@@ -372,16 +374,6 @@
 		return arguments.get(index);
 	}
 
-	private Optional<TypeInformation<?>> getPropertyInformation(String fieldname) {
-
-		var rawType = getType();
-		var field = ReflectionUtils.findField(rawType, fieldname);
-
-		return field != null ? Optional.of(TypeInformation.of(ResolvableType.forField(field, resolvableType)))
-				: Optional.ofNullable(BeanUtils.getPropertyDescriptor(rawType, fieldname))
-						.filter(it -> it.getName().equals(fieldname)).map(it -> from(it, rawType)).map(TypeInformation::of);
-	}
-
 	private ResolvableType from(PropertyDescriptor descriptor, Class<?> rawType) {
 
 		var method = descriptor.getReadMethod();
@@ -402,4 +394,27 @@
 	private boolean isNullableWrapper() {
 		return NullableWrapperConverters.supports(getType());
 	}
+
+	private Map<String, TypeInformation<?>> doGetProperties() {
+
+		Map<String, TypeInformation<?>> result = new HashMap<>();
+		Class<?> type = getType();
+		FieldCallback callback = field -> result.put(field.getName(),
+				TypeInformation.of(ResolvableType.forField(field, resolvableType)));
+
+		// Inspect fields first
+		ReflectionUtils.doWithFields(type, callback);
+
+		// Also inspect property descriptors
+		for (PropertyDescriptor descriptor : BeanUtils.getPropertyDescriptors(type)) {
+
+			if (result.containsKey(descriptor.getName())) {
+				continue;
+			}
+
+			result.put(descriptor.getName(), TypeInformation.of(from(descriptor, type)));
+		}
+
+		return Collections.unmodifiableMap(result);
+	}
 }
--- a/src/main/java/org/springframework/data/web/MapDataBinder.java
+++ b/src/main/java/org/springframework/data/web/MapDataBinder.java
@@ -52,25 +52,41 @@
  *
  * @author Oliver Gierke
  * @author Johannes Englmeier
+ * @author Christoph Strobl
  * @since 1.10
  */
 class MapDataBinder extends WebDataBinder {
 
+	static final int DEFAULT_COLLECTION_LIMIT = 1024;
+
 	private final Class<?> type;
 	private final ConversionService conversionService;
 
 	/**
-	 * Creates a new {@link MapDataBinder} for the given type and {@link ConversionService}.
+	 * Creates a new {@link MapDataBinder} for the given type and {@link ConversionService},
+	 * limiting collections to a size of {@link #DEFAULT_COLLECTION_LIMIT} for data binding.
 	 *
 	 * @param type target type to detect property that need to be bound.
 	 * @param conversionService the {@link ConversionService} to be used to preprocess values.
 	 */
 	public MapDataBinder(Class<?> type, ConversionService conversionService) {
+		this(type, conversionService, DEFAULT_COLLECTION_LIMIT);
+	}
 
+	/**
+	 * Creates a new {@link MapDataBinder} for the given type and {@link ConversionService}.
+	 *
+	 * @param type target type to detect property that need to be bound.
+	 * @param conversionService the {@link ConversionService} to be used to preprocess values.
+	 * @param collectionSizeLimit the maximum size of a collection to be bound.
+	 * @since 3.5.12
+	 */
+	public MapDataBinder(Class<?> type, ConversionService conversionService, int collectionSizeLimit) {
 		super(new HashMap<String, Object>());
 
 		this.type = type;
 		this.conversionService = conversionService;
+		this.setAutoGrowCollectionLimit(collectionSizeLimit);
 	}
 
 	@NonNull
@@ -89,7 +105,8 @@
 
 	@Override
 	protected ConfigurablePropertyAccessor getPropertyAccessor() {
-		return new MapPropertyAccessor(type, getTarget(), conversionService);
+		return new MapPropertyAccessor(type, getTarget(), conversionService, true,
+			getAutoGrowCollectionLimit());
 	}
 
 	/**
@@ -98,22 +115,32 @@
 	 *
 	 * @author Oliver Gierke
 	 * @author Johannes Englmeier
+	 * @author Christoph Strobl
 	 * @since 1.10
 	 */
 	private static class MapPropertyAccessor extends AbstractPropertyAccessor {
 
 		private static final SpelExpressionParser PARSER = new SpelExpressionParser(
-				new SpelParserConfiguration(false, true));
+			new SpelParserConfiguration(false, true, MapDataBinder.DEFAULT_COLLECTION_LIMIT));
 
 		private final Class<?> type;
 		private final Map<String, Object> map;
 		private final ConversionService conversionService;
+		private final SpelExpressionParser parser;
 
-		public MapPropertyAccessor(Class<?> type, Map<String, Object> map, ConversionService conversionService) {
+		public MapPropertyAccessor(Class<?> type, Map<String, Object> map, ConversionService conversionService,
+			boolean autoGrowCollections, int maxAutoGrowCollectionSize) {
 
 			this.type = type;
 			this.map = map;
 			this.conversionService = conversionService;
+
+			if (!autoGrowCollections || maxAutoGrowCollectionSize != MapDataBinder.DEFAULT_COLLECTION_LIMIT) {
+				parser = new SpelExpressionParser(
+					new SpelParserConfiguration(false, autoGrowCollections, maxAutoGrowCollectionSize));
+			} else {
+				parser = PARSER;
+			}
 		}
 
 		@Override
@@ -180,10 +207,11 @@
 			EvaluationContext context = SimpleEvaluationContext //
 					.forPropertyAccessors(new PropertyTraversingMapAccessor(type, conversionService)) //
 					.withConversionService(conversionService) //
+					.withAssignmentDisabled() //
 					.withRootObject(map) //
 					.build();
 
-			Expression expression = PARSER.parseExpression(propertyName);
+			Expression expression = parser.parseExpression(propertyName);
 
 			try {
 				expression.setValue(context, value);
--- a/src/main/java/org/springframework/data/web/ProxyingHandlerMethodArgumentResolver.java
+++ b/src/main/java/org/springframework/data/web/ProxyingHandlerMethodArgumentResolver.java
@@ -26,10 +26,13 @@
 import org.springframework.beans.factory.BeanFactoryAware;
 import org.springframework.beans.factory.ObjectFactory;
 import org.springframework.core.MethodParameter;
+import org.springframework.core.SpringProperties;
 import org.springframework.core.annotation.AnnotatedElementUtils;
 import org.springframework.core.convert.ConversionService;
 import org.springframework.data.projection.SpelAwareProxyProjectionFactory;
 import org.springframework.util.ClassUtils;
+import org.springframework.util.NumberUtils;
+import org.springframework.util.StringUtils;
 import org.springframework.web.bind.WebDataBinder;
 import org.springframework.web.bind.annotation.ModelAttribute;
 import org.springframework.web.bind.support.WebDataBinderFactory;
@@ -46,10 +49,13 @@
 public class ProxyingHandlerMethodArgumentResolver extends ModelAttributeMethodProcessor
 		implements BeanFactoryAware, BeanClassLoaderAware {
 
+	public static final String COLLECTION_SIZE_LIMIT_PARAM = "spring.data.web.projection.collection-limit";
+
 	private static final List<String> IGNORED_PACKAGES = Arrays.asList("java", "org.springframework");
 
 	private final SpelAwareProxyProjectionFactory proxyFactory;
 	private final ObjectFactory<ConversionService> conversionService;
+	private final int collectionSizeLimit;
 
 	/**
 	 * Creates a new {@link PageableHandlerMethodArgumentResolver} using the given {@link ConversionService}.
@@ -63,6 +69,13 @@
 
 		this.proxyFactory = new SpelAwareProxyProjectionFactory();
 		this.conversionService = conversionService;
+
+		String sizeFromProperty = SpringProperties.getProperty(COLLECTION_SIZE_LIMIT_PARAM);
+		if(StringUtils.hasText(sizeFromProperty)) {
+			this.collectionSizeLimit = NumberUtils.parseNumber(sizeFromProperty, Integer.class);
+		} else {
+			this.collectionSizeLimit = MapDataBinder.DEFAULT_COLLECTION_LIMIT;
+		}
 	}
 
 	@Override
@@ -118,7 +131,7 @@
 	protected Object createAttribute(String attributeName, MethodParameter parameter, WebDataBinderFactory binderFactory,
 			NativeWebRequest request) throws Exception {
 
-		MapDataBinder binder = new MapDataBinder(parameter.getParameterType(), conversionService.getObject());
+		MapDataBinder binder = new MapDataBinder(parameter.getParameterType(), conversionService.getObject(), this.collectionSizeLimit);
 		binder.bind(new MutablePropertyValues(request.getParameterMap()));
 
 		return proxyFactory.createProjection(parameter.getParameterType(), binder.getTarget());
--- a/src/test/java/org/springframework/data/mapping/PropertyPathUnitTests.java
+++ b/src/test/java/org/springframework/data/mapping/PropertyPathUnitTests.java
@@ -439,6 +439,21 @@
 				.isThrownBy(() -> PropertyPath.from(path, Left.class));
 	}
 
+	@Test // DATACMNS-1285, GH-3494
+	void rejectsTooLongPropertyNameSplit() {
+
+		String source = "AaBbCcDdeEfFgGhHiIjJkKlL";
+
+		for (int i = 0; i < 9; i++) {
+			source += source;
+		}
+
+		String path = source;
+
+		assertThatIllegalArgumentException() //
+			.isThrownBy(() -> PropertyPath.from(path, Left.class));
+	}
+
 	@Test // DATACMNS-1304
 	void resolvesPropertyPathWithSingleUppercaseLetterPropertyEnding() {
 		assertThat(from("categoryB", Product.class).toDotPath()).isEqualTo("categoryB");
--- a/src/test/java/org/springframework/data/web/MapDataBinderUnitTests.java
+++ b/src/test/java/org/springframework/data/web/MapDataBinderUnitTests.java
@@ -40,6 +40,7 @@
  * Unit tests for {@link MapDataBinder}.
  *
  * @author Oliver Gierke
+ * @author Christoph Strobl
  */
 class MapDataBinderUnitTests {
 
@@ -118,6 +119,31 @@
 				.withCauseInstanceOf(SpelEvaluationException.class);
 	}
 
+	@Test // GH-3495
+	void acceptsValuesWithinConfiguredMaxCollectionSize() {
+
+		MapDataBinder dataBinder = new MapDataBinder(Bar.class, new DefaultFormattingConversionService());
+		dataBinder.setAutoGrowCollectionLimit(300);
+		var accessor = dataBinder.getPropertyAccessor();
+
+		accessor.setPropertyValue("fooBar[299]", "Spring");
+
+		assertThat((Collection<?>) dataBinder.getTarget().get("fooBar")).hasSize(300);
+	}
+
+	@Test // GH-3495
+	void rejectsValuesBeyondConfiguredMaxCollectionSize() {
+
+		var accessor = new MapDataBinder(Bar.class, new DefaultFormattingConversionService())
+			.getPropertyAccessor();
+
+		assertThatNoException().isThrownBy(() -> accessor.setPropertyValue("fooBar[299]", "Spring"));
+
+		assertThatExceptionOfType(NotWritablePropertyException.class) //
+			.isThrownBy(() -> accessor.setPropertyValue("fooBar[2000]", "Spring")) //
+			.withCauseInstanceOf(SpelEvaluationException.class);
+	}
+
 	private static Map<String, Object> bind(PropertyValues values) {
 
 		var binder = new MapDataBinder(Root.class, new DefaultFormattingConversionService());
